Indicate deprecated and unused OpenSSL errors

valcikeric opened the pull request:
Hi,

Here is my proposal to fix Issue #24, feel free to tell me what to improve.

I separated the X509_V_ERR_OUT_OF_MEM into a commit of its own (1441c3e) because I'm not really sure if this error qualifies as unused.

mukrop commented on web/_includes/icon.html:
```diff
@@ -23,6 +23,8 @@
     <span class="fas fa-fw fa-project-diagram color-white"></span>
   {% when 'button-email' %}
     <span class="fas fa-fw fa-envelope color-white"></span>
+  {% when 'unused-openssl-icon' %}
```
The key has a bit unintuitive name. When you look on the other icon keys, none has the word "icon" and if it's bound to a library, the library name is a prefix. I thus suggest `openssl-unused` instead.

mukrop replied:
Furthermore, please move it higher in the file right after the library icons.

mukrop commented on web/_includes/icon.html:
```diff
@@ -23,6 +23,8 @@
   {% when 'button-email' %}
     <span class="fas fa-fw fa-envelope color-white"></span>
+  {% when 'unused-openssl-icon' %}
+    <span class="fas fa-times {{ include.float }} color-verylightgray"></span>
```
Please use a fixed-width icon to be consistent with the style (adding the class `fa-fw`).

mukrop replied:
Secondly, add a hover title text with the explanation (see the library icons).

mukrop commented on errors/X509_V_ERR_UNSPECIFIED/data.yml:
```diff
@@ -1,4 +1,5 @@
 tags: [ openssl, uncategorized ]
 openssl-docs: |
   Unspecified error; should not happen.
+unused-openssl: |
```
The key name is inconsistent with the others. we use the library name as a prefix, not a suffix (see `openssl-docs`, `gnutls-docs` and similar). I thus suggest `openssl-unused`.
Secondly, the empty multiline string denoted by the pipe (`|`) seems a bit counterintuitive. I would prefer `openssl-unused: true`.

mukrop commented on web/_pages/index.html:
```diff
@@ -8,7 +8,7 @@ <h1 id="home">Making X.509 errors usable.</h1>
     <p>Validating X.509 certificates correctly turns out to be pretty complicated (e.g. <a href="http://www.cs.utexas.edu/~shmat/shmat_ccs12.pdf">Georgiev2012</a>, <a href="https://crocs.fi.muni.cz/papers/acsac2019">Ukrop2019</a>). Yet certificate validation is absolutely crucial for secure communication on the Internet (think <a href="https://en.wikipedia.org/wiki/Transport_Layer_Security">TLS</ ...
     <p>Our goal is to simplify the ecosystem by consolidating the errors and their documentation (similarly to <a href="https://blog.mozilla.org/blog/2017/10/18/mozilla-brings-microsoft-google-w3c-samsung-together-create-cross-browser-documentation-mdn/">web documentation</a>) and by explaining better what the validation errors mean.</p>
-    <p>For every error, we aim to provide an example certificate ({% include icon.html icon="certificate" %}), documentation from OpenSSL ({% include icon.html icon="openssl-book" %}) and other TLS libraries ({% include icon.html icon="gnutls-book" %}, {% include icon.html icon="botan-book" %}, {% include icon.html icon="mbed-book" %}). In the future, we plan the possibility of reorganization bas ...
+    <p>For every error, we aim to provide an example certificate ({% include icon.html icon="certificate" %}), documentation from OpenSSL ({% include icon.html icon="openssl-book" %}, followed by {% include icon.html icon="unused-openssl-icon" %} if that error is not used anymore) and other TLS libraries ({% include icon.html icon="gnutls-book" %}, {% include icon.html icon="botan-book" %}, {% in ...
```
Let's be a bit more concise and precise in the parenthesis: "<book>, deprecated or unused denoted by <cross>".

mukrop replied:
Second note: Be precise in the naming. Is it unused errors (nowhere in the code), deprecated errors (still present in the code) or both? (I think it's both.)

valcikeric commented:
Thank you for the feedback.
I fixed all the issues mentioned above. Also, I'm now aware that I labeled these error codes without much thought. So I did a little research and divided the 8 codes into 3 groups, each group has its own commit.

__X509_V_ERR_OUT_OF_MEMORY__, __X509_V_ERR_UNSPECIFIED__
- On second thought, these codes are not deprecated or unused in my opinion, so I removed the `openssl-unused`. 
- commit f312406

__X509_V_ERR_APPLICATION_VERIFICATION__, __X509_V_ERR_UNABLE_TO_DECRYPT_CRL_SIGNATURE__
- These codes are mentioned in the code but are not returned. (I could be wrong)
- commit 8c71ffe

__X509_V_ERR_AKID_SKID_MISMATCH__, __X509_V_ERR_AKID_ISSUER_SERIAL_MISMATCH__, __X509_V_ERR_SUBJECT_ISSUER_MISMATCH__, __X509_V_ERR_KEYUSAGE_NO_CERTSIGN__
- These codes are not only mentioned in the code but also returned by some functions. I did some tracking by hand and I believe that these codes cannot be returned all the way to the user. (again, I could be easily wrong)
- These codes can be returned all the way to a function `check_issued` in `x509_vfy.c`, but this function only returns 1 or 0, so the error codes won't be propagated deeper.
- commit b98463f

zacikpa commented:
> 
> **X509_V_ERR_OUT_OF_MEMORY**, **X509_V_ERR_UNSPECIFIED**
> 
> * On second thought, these codes are not deprecated or unused in my opinion, so I removed the `openssl-unused` tag.

Indeed, these two are neither deprecated nor unused.
I would leave them as they are (only uncategorized) since they have nothing to do with the actual path validation.
> 
> **X509_V_ERR_APPLICATION_VERIFICATION**, **X509_V_ERR_UNABLE_TO_DECRYPT_CRL_SIGNATURE**
> 
> * These codes are mentioned in the code but are not returned (I could be wrong). So I've left them labeled as unused (deprecated).

**X509_V_ERR_APPLICATION_VERIFICATION** 
This is a very specific error that applications using the openssl API might return. 
I understand its meaning like this: While the library cannot produce it by itself, it is the recommended error to use when something else in your application fails and you want to return an openssl-compatible error.
I would leave this one *without* the unused tag.

**X509_V_ERR_UNABLE_TO_DECRYPT_CRL_SIGNATURE**
This one truly only has a definition in the code, the *unused* tag suits it well.
> 
> **X509_V_ERR_AKID_SKID_MISMATCH**, **X509_V_ERR_AKID_ISSUER_SERIAL_MISMATCH**, **X509_V_ERR_SUBJECT_ISSUER_MISMATCH**, **X509_V_ERR_KEYUSAGE_NO_CERTSIGN**
> 
> * These codes are not only mentioned in the code but also returned by some functions. I did some tracking by hand and I believe that these codes cannot be returned all the way to the user. (again, I could be easily wrong)
> * These codes can be returned all the way to a function `check_issued` in `x509_vfy.c`, but this function only returns 1 or 0, so the error codes won't be propagated deeper. So again, I left them labeled as unused.

Good job tracking that down. These four are also marked *deprecated* in the official docs, so your assumptions are correct.

zacikpa commented:
There are still at least three errors that I would personally tag unused.
If you are interested, take a deeper look at:
- X509_V_ERR_UNABLE_TO_DECRYPT_CRL_SIGNATURE
- X509_V_ERR_UNSUPPORTED_EXTENSION_FEATURE
- X509_V_ERR_UNSUPPORTED_CONSTRAINT_SYNTAX

Though I am pretty sure that we can mark these three unused right away.


I am not really sure what to do with these:
- X509_V_ERR_OCSP_VERIFY_NEEDED
- X509_V_ERR_OCSP_VERIFY_FAILED
- X509_V_ERR_OCSP_CERT_UNKNOWN

But I assume they might be returned by higher-level applications in the same way as X_509_V_ERR_APPLICATION_VERIFICATION, so I wouldn't personally tag them.

Thank you for spending your time on this.

valcikeric commented:
I'll take a look at those three you mentioned and add them to the rest.
I will also remove the label `openssl-unused` from X509_V_ERR_APPLICATION_VERIFICATION.

Thank you for your input.

valcikeric commented:
I now pushed the last changes I wanted to make.

I removed the `openssl-unused` tag from __X509_V_ERR_APPLICATION_VERIFICATION__ and added it to __X509_V_ERR_UNSUPPORTED_EXTENSION_FEATURE__ and __X509_V_ERR_UNSUPPORTED_CONSTRAINT_SYNTAX__.

I already had error code __X509_V_ERR_UNABLE_TO_DECRYPT_CRL_SIGNATURE__ tagged as unused or deprecated, so I left it as is.

valcikeric commented:
I've made the requested changed.

I also added two new codes as deprecated/unused: __X509_V_ERR_PROXY_SUBJECT_INVALID__ and __X509_V_ERR_PATH_LOOP__

__X509_V_ERR_PROXY_SUBJECT_INVALID__ is not even defined in the code.
__X509_V_ERR_PATH_LOOP__ - I wanted to replicate this code, but I failed. I believe that this code cannot be returned from OpenSSL. It is only used in function `check_issued` in `x509_vfy.c`, but this function returns only 1 or 0. Also, there is no way of propagating this error outside this function (e.g. via pointers). It is only assigned to variable `ret` and this variable ceases to exist outside of this function.

valcikeric commented:
Ah, in the last commit I changed the `sed` command in Makefile in the __X509_V_ERR_CERT_SIGNATURE_FAILURE__ folder in hopes of fixing why the Travis check failed. Travis passed on my repo or another branch, but here it failed again. I can remove the changed `sed` command to the one that was there before.

I used the `sed` command to not match the pattern only on single lines but to always add the following line to the current compared line and after that try to match the patter. That way I thought I could identify the last line of the certificate and change the last couple of chars to make the signature invalid. I found this solution [here](https://stackoverflow.com/a/1252191) and modified it a bit.

zacikpa commented:
In the not yet merged #66, I've replaced the signature failure certificate with a static one, so that it doesn't cause any issues in the future.
If you come up with a better solution, feel free to change it, but I'd suggest making a separate PR for that.

zacikpa commented:
> @zacikpa: Should I just merge it anyway and you'll rebase #66 on the new master and rework it into a static chain for the CI to pass?

I'm okay with that.
